### PR TITLE
fix: restore each rank's own RNG state on checkpoint load

### PR DESCRIPTION
**Pull request by ZhiyuLi-Nvidia**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Fixes `load_checkpoint` restoring the wrong RNG state on most ranks.

## Issue tracking

Linked issue: 

## The bug

`cuda_rng_tracker` streams are seeded **per rank** so ranks draw independently
(`tensor_parallel/random.py`: `expert_parallel_seed = seed + 1024 + 100*ep_rank + etp_rank`).

Those streams are consumed on every forward. The MoE router draws from this one
(`moe_utils.py:1277`):

```python
with get_cuda_rng_tracker().fork(get_expert_parallel_rng_tracker_name()):
    random_logits = logits.clone().normal_()      # advances the expert-parallel stream
```

Expert dropout forks the same tracker (`extensions/transformer_engine.py:1242`).

But `get_rng_state` passed `replica_id=dp_cp_rank`, which tells dist-checkpointing *"my copy
duplicates someone else's"*. That is false. Only one rank's state was written, and its peers
restored it — so any rank with `ep_rank > 0` resumed on someone else's stream, and the independence
the seeding exists to create was gone for the rest of the run.

Measured on a 120B MoE run at 64 nodes: **`expert-parallel-rng` differed from the saved value on
252 of 256 ranks**; every correct rank was an expert-parallel rank 0. Not gated on any flag — any
resume with EP > 1 hits it, and the load reports success.

## The fix

Move `dp_cp_rank` out of `replica_id` and into the offset. These ranks hold different data, so they
need different addresses:

```python
-  (pp_size, tp_size),              (pp_rank, tp_rank),  replica_id=dp_cp_rank
+  (pp_size, tp_size, dp_cp_size),  (pp_rank, tp_rank, dp_cp_rank)
```

## What reviewers should know

- **Old checkpoints won't load their RNG state** — 2-D key vs 3-D request. This is the
  compatibility break, and the operator sees a raw dist-checkpointing key error rather than a
  graceful skip; `--no-load-rng` is the workaround. Recording the key layout in the checkpoint's
  content metadata would let the load path degrade the way the world-size case does — worth doing,
  but a larger change than this fix.
- **DP rescale** now keeps its freshly-seeded RNG and logs why, rather than failing
  (`load_checkpoint` gained a `world_size` check beside the existing `(TP, PP)` one).
- **Size**: +47 MB at 4096 ranks. The old count was smaller only because it discarded most ranks'
  state. No extra files, no added collectives.
- **Scope**: `torch_dist` only. `fsdp_dtensor` has the same defect, marked `TODO` in source.

`dp_cp` separates more ranks than strictly necessary — `dp`/`cp` enter no seed, so ranks sharing an
`ep_rank` hold identical state. The tighter key `(pp, tp, ep, etp, …)` is rejected by the format:
`ShardedObject` needs a product-shaped `global_shape` and those axes are interdependent. `dp_cp` is
the least-separating legal key.

The third commit is separate and smaller: `DataLoader` iterator creation draws from the default CPU
generator *after* `load_checkpoint` restores it, so `torch_rng_state` doesn't round-trip. It does
not change training output — resume position comes from `consumed_samples`, and samplers shuffle
with their own epoch-seeded generators.

## Testing

Every test was also run against `main` — one that passes either way proves nothing.

| test | this PR | `main` |
|---|---|---|
| RNG round trip, `(tp,pp,cp,ep)` = (1,1,1,2), (2,1,1,2), (1,1,2,2) | pass | **fail** — `restored a different 'expert-parallel-rng' stream` |
| resumed loader continues a saved run (same samples + same CPU RNG state) | pass | **fail** — `the resumed process holds a different CPU RNG state than the run it continued` |

The `cp=2` case matters: a `(pp, tp, dp)` key passes at `cp=1` and collides differing ranks at
`cp=2`.

End-to-end, 2 nodes EP=4, checkpoint at step 5 of 10, comparing the resumed run's checkpoint
against a continuous run's byte-for-byte — clean with the router flag on, with it off, and with
`--data-parallel-random-init`.

At 64 nodes, EP=64, resume at step 25 of 50: continuous and resumed match on lm loss, grad norm and
params norm at full float32, while the same comparison still reports the pre-fix run diverging at
step 26 (`grad-norm rel 2.8e-05`).

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### CODE REVIEW

**Review by yaoyu-33 — APPROVED** [review]

**Review by Connor-XY — APPROVED** [review]

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6858/commits/13c0355d58b970ce2f1d9c2185e1a11220cd9b41

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6858/commits/a1dad480cae590d350e66d2580d79b43906d5fc4

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6858/commits/2071e6f6ac5ab4c37746d9a422e966debeeeff44

**Review by jaredcasper — APPROVED** [review]

**Comment by ZhiyuLi-Nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6858/commits/88c3d77a91915ee9550f9b7b7af1133493d0de5c
